README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
detector
========

[![Build Status](https://travis-ci.org/paulhendricks/detector.png?branch=master)](https://travis-ci.org/paulhendricks/detector) [![Build status](https://ci.appveyor.com/api/projects/status/gu5ggnr1i2muw5r3/branch/master?svg=true)](https://ci.appveyor.com/project/paulhendricks/detector/branch/master) [![codecov.io](http://codecov.io/github/paulhendricks/detector/coverage.svg?branch=master)](http://codecov.io/github/paulhendricks/detector?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/detector)](http://cran.r-project.org/package=detector) [![Downloads from the RStudio CRAN mirror](http://cranlogs.r-pkg.org/badges/detector)](http://cran.rstudio.com/package=detector) [![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active)

`detector` makes detecting data containing [Personally Identifiable Information](https://en.wikipedia.org/wiki/Personally_identifiable_information) (PII) quick, easy, and scalable. It provides high-level functions that can take vectors and data.frames and return important summary statistics in a convenient data.frame. Once complete, `detector` will be able to detect the following types of PII:

-   Full name
-   Home address
-   E-mail address
-   National identification number
-   Passport number
-   Social Security number
-   IP address
-   Vehicle registration plate number
-   Driver's license number
-   Credit card number
-   Date of birth
-   Birthplace
-   Telephone number
-   Latitude and longtiude

State of the Union
------------------

### Complete!

-   E-mail address
-   Telephone number
-   National identification number

### Needs more work...

-   Credit card number

### Haven't even started :(

-   Full name
-   Date of birth
-   Home address
-   IP address
-   Vehicle registration plate number
-   Driver's license number
-   Birthplace
-   Latitude and longtiude

Installation
------------

You can install:

-   the latest released version from CRAN with

    ``` r
    install.packages("detector")
    ```

-   the latest development version from github with

    ``` r
    if (packageVersion("devtools") < 1.6) {
      install.packages("devtools")
    }
    devtools::install_github("paulhendricks/detector")
    ```

If you encounter a clear bug, please file a minimal reproducible example on [github](https://github.com/paulhendricks/detector/issues).

API
---

### Generate data containing fake PII

``` r
library(dplyr, warn.conflicts = FALSE)
library(generator)
n <- 6
set.seed(1)
ashley_madison <- 
  data.frame(name = r_full_names(n), 
             snn = r_national_identification_numbers(n), 
             dob = r_date_of_births(n), 
             email = r_email_addresses(n), 
             ip = r_ipv4_addresses(n), 
             phone = r_phone_numbers(n), 
             credit_card = r_credit_card_numbers(n), 
             lat = r_latitudes(n), 
             lon = r_longitudes(n), 
             stringsAsFactors = FALSE)
knitr::kable(ashley_madison, format = "markdown")
```

| name                  | snn         | dob        | email                    | ip              | phone      | credit\_card        |          lat|          lon|
|:----------------------|:------------|:-----------|:-------------------------|:----------------|:-----------|:--------------------|------------:|------------:|
| Eldridge Pfannerstill | 442-34-5338 | 1991-11-11 | <ntakqojv@lgbcyk.rkv>    | 45.84.71.225    | 6794976958 | 4125-7204-9193-5140 |   -2.7018575|     8.634988|
| Augustine Homenick    | 799-44-6396 | 1912-06-27 | <iqg@mtcuh.viy>          | 191.116.55.106  | 3275827694 | 2182-5994-2283-9486 |  -70.4148630|   -65.827918|
| Jennie Runte          | 941-11-5441 | 1983-09-13 | <wjszy@sjhreocvt.gbp>    | 27.128.73.17    | 7419351735 | 4370-4866-4735-7857 |  -45.4091701|   -79.932229|
| Araceli Kunde         | 290-44-2675 | 1947-07-26 | <uljsnvhfr@qfdkumtn.jkd> | 221.47.229.86   | 3243246285 | 6682-5074-2898-9396 |   -0.2673845|   103.514583|
| Josue Rau             | 686-88-8446 | 1994-12-10 | <c@lqxzkdpi.nfy>         | 157.136.114.185 | 9169736873 | 4510-3757-4858-5236 |  -22.8839925|    72.886505|
| Elnora Zemlak         | 212-40-7016 | 1974-10-30 | <capvnl@nympzf.gsk>      | 143.20.199.87   | 3295843196 | 7206-6205-2194-6432 |   78.2444466|  -120.590050|

### Detect data containing PII

``` r
library(detector)
ashley_madison %>% 
  detect %>% 
  knitr::kable(format = "markdown")
```

| column\_name | has\_email\_addresses | has\_phone\_numbers | has\_national\_identification\_numbers |
|:-------------|:----------------------|:--------------------|:---------------------------------------|
| name         | FALSE                 | FALSE               | FALSE                                  |
| snn          | FALSE                 | FALSE               | TRUE                                   |
| dob          | FALSE                 | FALSE               | FALSE                                  |
| email        | TRUE                  | FALSE               | FALSE                                  |
| ip           | FALSE                 | FALSE               | FALSE                                  |
| phone        | FALSE                 | TRUE                | FALSE                                  |
| credit\_card | FALSE                 | FALSE               | FALSE                                  |
| lat          | FALSE                 | TRUE                | FALSE                                  |
| lon          | FALSE                 | TRUE                | FALSE                                  |


NEWS.md
# detector 0.1.0.9000

## Improvements

* TO BE EDITED.

# detector 0.1.0

## Improvements

* Added functions for e-mail addresses, phone numbers, and national identification numbers.


tests/testthat.R
library(testthat)
library(detector)

test_check("detector")


tests/testthat/test-detect.R
library(generator)
library(detector)
context("detect()")

test_df <-
  data.frame(phone_number = r_phone_numbers(1000),
             email_address = r_email_addresses(1000),
             national_identification_number = r_national_identification_numbers(1000),
             mixed = c(r_phone_numbers(300), r_email_addresses(200), r_national_identification_numbers(500)),
             stringsAsFactors = FALSE)

test_that("Produces the correct output.", {
})

test_that("Produces the correct output type.", {
  expect_is(detect(letters), "data.frame")
  expect_is(detect(as.Date("2014-01-01")), "data.frame")
  expect_is(detect(test_df), "data.frame")
})

test_that("Produces the correct messages, warnings, and errors.", {
})



tests/testthat/test-phone_numbers.R
library(generator)
library(detector)
context("phone numbers")

test_that("Produces the correct output.", {
  expect_equal(is_phone_number("hello"), FALSE)
  expect_equal(is_phone_number(65884), FALSE)
  expect_equal(is_phone_number("111-333-5555"), TRUE)
  expect_equal(is_phone_number(1113335555), TRUE)
  expect_equal(has_phone_numbers("hello"), FALSE)
  expect_equal(has_phone_numbers(65884), FALSE)
  expect_equal(has_phone_numbers("111-333-5555"), TRUE)
  expect_equal(has_phone_numbers(1113335555), TRUE)
})

test_that("Produces the correct output type.", {
  expect_is(is_phone_number("hello"), "logical")
  expect_is(is_phone_number("hello@world.edu"), "logical")
  expect_is(is_phone_number("hello"), "logical")
  expect_is(is_phone_number(65884), "logical")
  expect_is(is_phone_number("111-333-5555"), "logical")
  expect_is(is_phone_number(1113335555), "logical")
  expect_is(has_phone_numbers("hello"), "logical")
  expect_is(has_phone_numbers("hello@world.edu"), "logical")
  expect_is(has_phone_numbers("hello"), "logical")
  expect_is(has_phone_numbers(65884), "logical")
  expect_is(has_phone_numbers("111-333-5555"), "logical")
  expect_is(has_phone_numbers(1113335555), "logical")
})

test_that("Produces the correct errors.", {
})


R/detect.R
#' Detect if a data object contains PII.
#'
#' @param .x A data object.
#' @return A logical value indicating if that data object contains PII.
#' @examples
#' # atomic vectors
#' detect(letters)
#' detect(1:10)
#' detect(as.Date("2014-01-01"))
#'
#' # data.frames
#' detect(mtcars)
#' @export
detect <- function(.x) {
  UseMethod("detect")
}

#' @describeIn detect Method for default vectors.
#' @export
detect.default <- function(.x) {
  return(detect(as.character(.x)))
}

#' @describeIn detect Method for character vectors.
#' @export
detect.character <- function(.x) {
  temp <- data.frame()
  temp[1, "has_email_addresses"] <- has_email_addresses(.x)
  temp[, "has_phone_numbers"] <- has_phone_numbers(.x)
  temp[, "has_national_identification_numbers"] <- has_national_identification_numbers(.x)
  return(temp)
}

detect_column <- function(.nm, .df) {
  return(cbind(data.frame(column_name = .nm, stringsAsFactors = FALSE), detect(.df[, .nm])))
}

#' @describeIn detect Method for data.frames.
#' @export
detect.data.frame <- function(.x) {
  temp <- lapply(colnames(.x), detect_column, .df = .x)
  return(do.call(rbind, temp))

}


R/email_addresses.R
#' Test if a string is an e-mail address.
#'
#' @param .x A character vector.
#' @return A logical value indicating if that string is an e-mail address.
#' @examples
#' # Examples
#' is_email_address("hello") # FALSE
#' is_email_address("hello@@world.edu") # TRUE
#' @export
is_email_address <- function(.x) {
  return(stringr::str_detect(stringr::str_trim(as.character(.x)),
                             "^[[:alnum:].-]+@[[:alnum:].-]+$"))
}

#' Test if a character vector has any e-mail addresses.
#'
#' @param .x A character vector.
#' @return A logical value indicating if that string has any e-mail addresses.
#' @examples
#' # Examples
#' has_email_addresses("hello") # FALSE
#' has_email_addresses("hello@@world.edu") # TRUE
#' @export
has_email_addresses <- function(.x) {
  return(any(is_email_address(.x)))
}


R/national_identification_numbers.R
#' Test if a string is a national identification number.
#'
#' Strictly works for only US national identification numbers.
#'
#' @param .x A string or numeric vector.
#' @return A logical value indicating if that string is a national identification number.
#' @examples
#' # Examples
#' is_national_identification_number("hello") # FALSE
#' is_national_identification_number(65884) # FALSE
#' is_national_identification_number("111-33-5555") # TRUE
#' is_national_identification_number(1113335555) # FALSE
#' @export
is_national_identification_number <- function(.x) {
  return(stringr::str_detect(stringr::str_trim(as.character(.x)),
                             "^[0-9]{3}-[0-9]{2}-[0-9]{4}$"))
}

#' Test if a character vector has any national identification numbers.
#'
#' @param .x A character vector.
#' @return A logical value indicating if that string has any national identification numbers.
#' @examples
#' # Examples
#' # Examples
#' has_national_identification_numbers("hello") # FALSE
#' has_national_identification_numbers(65884) # FALSE
#' has_national_identification_numbers("111-33-5555") # TRUE
#' has_national_identification_numbers(1113335555) # FALSE
#' @export
has_national_identification_numbers <- function(.x) {
  return(any(is_national_identification_number(.x)))
}


R/phone_numbers.R
#' Test if a string is a phone number.
#'
#' Strictly works for only US phone numbers.
#'
#' @param .x A string or numeric vector.
#' @return A logical value indicating if that string is a phone number.
#' @examples
#' # Examples
#' is_phone_number("hello") # FALSE
#' is_phone_number(65884) # FALSE
#' is_phone_number("111-333-5555") # TRUE
#' is_phone_number(1113335555) # TRUE
#' @export
is_phone_number <- function(.x) {
  return(stringr::str_detect(stringr::str_trim(as.character(.x)),
                             "\\d{3}?[.-]? *\\d{3}[.-]? *[.-]?\\d{4}"))
}

#' Test if a character vector has any phone numbers.
#'
#' @param .x A character vector.
#' @return A logical value indicating if that string has any phone numbers.
#' @examples
#' # Examples
#' has_phone_numbers("hello") # FALSE
#' has_phone_numbers(65884) # FALSE
#' has_phone_numbers("111-333-5555") # TRUE
#' has_phone_numbers(1113335555) # TRUE
#' @export
has_phone_numbers <- function(.x) {
  return(any(is_phone_number(.x)))
}
